### PR TITLE
Spring Cloud Azure App Configuration 4.0

### PR DESCRIPTION
**Pull request by mrm9084**
# Description
This updates the App Configuration Spring Provider to the 4.0 format, adds support for global properties

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [X] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [X] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/spring/spring-cloud-azure-appconfiguration-config-web/pom.xml
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-appconfiguration-config-web/pom.xml:51]: spring-boot-starter-test already has this dependency.

File: sdk/spring/spring-cloud-azure-appconfiguration-config-web/src/main/java/com/azure/spring/cloud/config/web/implementation/pullrefresh/AppConfigurationEventListener.java
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-appconfiguration-config-web/src/main/java/com/azure/spring/cloud/config/web/implementation/pullrefresh/AppConfigurationEventListener.java:3]: do you think `refresh.pull` and `refresh.push` would be better as the package names?

File: sdk/spring/spring-cloud-azure-appconfiguration-config-web/src/main/java/com/azure/spring/cloud/config/web/package-info.java
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-appconfiguration-config-web/src/main/java/com/azure/spring/cloud/config/web/package-info.java:6]: No need for this class, since there's no public classes.

File: sdk/spring/spring-cloud-azure-appconfiguration-config/pom.xml
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-appconfiguration-config/pom.xml:85]: Why do we need this?

File: sdk/spring/spring-cloud-azure-appconfiguration-config/pom.xml
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-appconfiguration-config/pom.xml:28]: What's this for?

File: sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/config/KeyVaultSecretProvider.java
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/config/KeyVaultSecretProvider.java:8]: We still need this?

**Review by saragluna — CHANGES_REQUESTED** [review]

File: eng/code-quality-reports/src/main/resources/revapi/revapi.json
Comment (mrm9084) [reply eng/code-quality-reports/src/main/resources/revapi/revapi.json:749]: we should avoid regex as it increases the revapi time. See Alan's post about it.

File: eng/versioning/version_client.txt
Comment (mrm9084) [reply eng/versioning/version_client.txt:183]: We have already released a beta of the feature management library. It is the code in the feature branch. https://github.com/Azure/azure-sdk-for-java/tree/feature/azconfig-spring/DynamicFeature

File: sdk/appconfiguration/tests.yml
Comment (mrm9084) [reply sdk/appconfiguration/tests.yml:4]: Doesn't the SDK also use that file?

File: sdk/spring/spring-cloud-azure-appconfiguration-config-web/src/main/java/com/azure/spring/cloud/config/web/package-info.java
Comment (mrm9084) [reply sdk/spring/spring-cloud-azure-appconfiguration-config-web/src/main/java/com/azure/spring/cloud/config/web/package-info.java:6]: The build seems to fail if this file is removed.

File: sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/config/KeyVaultSecretProvider.java
Comment (mrm9084) [reply sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/config/KeyVaultSecretProvider.java:8]: This is for resolving key vault references without connecting to key vault. It has no "client", it is instead of.

File: sdk/appconfiguration/tests.yml
Comment (saragluna) [reply sdk/appconfiguration/tests.yml:4]: Not sure, check this out https://dev.azure.com/azure-sdk/internal/_build/results?buildId=2169438&view=logs&j=1bcb7d81-906d-5a53-3f29-b3b6382abd7c&t=26e36aca-4901-52a9-d557-4eb0b6782169&l=15.

File: sdk/spring/spring-cloud-azure-appconfiguration-config-web/src/main/java/com/azure/spring/cloud/config/web/package-info.java
Comment (saragluna) [reply sdk/spring/spring-cloud-azure-appconfiguration-config-web/src/main/java/com/azure/spring/cloud/config/web/package-info.java:6]: Let me try locally.

File: sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/config/KeyVaultSecretProvider.java
Comment (saragluna) [reply sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/config/KeyVaultSecretProvider.java:8]: But internally we will build the KV clients?

File: eng/code-quality-reports/src/main/resources/revapi/revapi.json
Comment (saragluna) [reply eng/code-quality-reports/src/main/resources/revapi/revapi.json:749]: I don't think this mean we should avoid regex,

> When adding many suppressions determine if they can be compacted with existing suppressions for that kind
- RevApi has regex suppressions
- Checkstyle uses <Or>
- Spotbugs has regex suppressions

But I checked the PR, he is using code to address the external APIs. See `eng/code-quality-reports/src/main/java/com/azure/tools/revapi/transforms/AzureSdkAllowedExternalApis.java` in https://github.com/Azure/azure-sdk-for-java/pull/33130.

File: sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/config/KeyVaultSecretProvider.java
Comment (mrm9084) [reply sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/config/KeyVaultSecretProvider.java:8]: So the way this works is before we check if a client even exists we pass the KV Reference to the KeyVaultSecretProvider, if it returns any non-Null value than that is returned as the secret. If nothing is returned or this isn't configured only then do we create a client or check if one has already been made.

File: sdk/spring/spring-cloud-azure-appconfiguration-config-web/src/main/java/com/azure/spring/cloud/config/web/AppConfigurationWebAutoConfiguration.java
Comment (saragluna) [reply sdk/spring/spring-cloud-azure-appconfiguration-config-web/src/main/java/com/azure/spring/cloud/config/web/AppConfigurationWebAutoConfiguration.java:54]: I also think we could remove these JavaDocs now.

File: sdk/spring/spring-cloud-azure-appconfiguration-config-web/src/main/java/com/azure/spring/cloud/config/web/implementation/pullrefresh/AppConfigurationEventListener.java
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-appconfiguration-config-web/src/main/java/com/azure/spring/cloud/config/web/implementation/pullrefresh/AppConfigurationEventListener.java:7]: Line 20, if it's already defined as a bean, then there's no need to add the `@Component` annotation here.

File: sdk/spring/spring-cloud-azure-appconfiguration-config-web/src/main/java/com/azure/spring/cloud/config/web/implementation/pushbusrefresh/AppConfigurationBusRefreshEventListener.java
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-appconfiguration-config-web/src/main/java/com/azure/spring/cloud/config/web/implementation/pushbusrefresh/AppConfigurationBusRefreshEventListener.java:3]: Line 15, if this is already defined as a bean, then there's no need to add the `@Component` annotation.

File: sdk/spring/spring-cloud-azure-appconfiguration-config-web/src/main/java/com/azure/spring/cloud/config/web/implementation/pushrefresh/AppConfigurationRefreshEventListener.java
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-appconfiguration-config-web/src/main/java/com/azure/spring/cloud/config/web/implementation/pushrefresh/AppConfigurationRefreshEventListener.java:3]: Line 15, no need of the `@Component` annotation.

File: sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/feature/filters/PercentageFilter.java
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/feature/filters/PercentageFilter.java:18]: Why using this scan-based annotation, can we use `@Bean` instead?

File: sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/feature/filters/TimeWindowFilter.java
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/feature/filters/TimeWindowFilter.java:24]: Same for this one, why use the `@Component`.

File: sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/implementation/models/package-info.java
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/implementation/models/package-info.java:6]: We can remove this file.

File: sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/implementation/package-info.java
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/manager/implementation/package-info.java:8]: We can remove this file.

File: sdk/spring/spring-cloud-azure-feature-management-web/src/main/java/com/azure/spring/cloud/feature/manager/FeatureHandler.java
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-feature-management-web/src/main/java/com/azure/spring/cloud/feature/manager/FeatureHandler.java:29]: Same here, no need for the `@Component`.

File: sdk/spring/spring-cloud-azure-feature-management-web/src/main/java/com/azure/spring/cloud/feature/manager/implementation/FeatureManagementWebConfiguration.java
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-feature-management-web/src/main/java/com/azure/spring/cloud/feature/manager/implementation/FeatureManagementWebConfiguration.java:54]: Remove `public` from bean definitions.

File: sdk/spring/spring-cloud-azure-feature-management-web/src/main/java/com/azure/spring/cloud/feature/manager/implementation/package-info.java
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-feature-management-web/src/main/java/com/azure/spring/cloud/feature/manager/implementation/package-info.java:3]: We can remove this file.

File: sdk/spring/spring-cloud-azure-feature-management-web/src/main/java/com/azure/spring/cloud/feature/manager/IDisabledFeaturesHandler.java
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-feature-management-web/src/main/java/com/azure/spring/cloud/feature/manager/IDisabledFeaturesHandler.java:19]: Do we need this `@Component` annotation?

File: sdk/spring/spring-cloud-azure-test-appconfiguration-config/pom.xml
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-test-appconfiguration-config/pom.xml:16]: How about naming this `spring-cloud-azure-integration-test-appconfiguration-config`

Comment (saragluna) [reply sdk/spring/spring-cloud-azure-test-appconfiguration-config/pom.xml:16]: Have we added this to the tests.yml?

**Review by saragluna — CHANGES_REQUESTED** [review]

File: sdk/spring/spring-cloud-azure-appconfiguration-config-web/src/main/java/com/azure/spring/cloud/config/web/AppConfigurationWebAutoConfiguration.java
Comment (mrm9084) [reply sdk/spring/spring-cloud-azure-appconfiguration-config-web/src/main/java/com/azure/spring/cloud/config/web/AppConfigurationWebAutoConfiguration.java:54]: The public removal was the change on this line.

File: sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/config/implementation/config/AppConfigurationBootstrapConfiguration.java
Comment (mrm9084) [reply sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/config/implementation/config/AppConfigurationBootstrapConfiguration.java:51]: This change seems to cause an issue with SpotBugs.

File: eng/code-quality-reports/src/main/resources/revapi/revapi.json
Comment (saragluna) [reply eng/code-quality-reports/src/main/resources/revapi/revapi.json:749]: We can wait until #33130 gets merged.

File: sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/config/AppConfigurationRefresh.java
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/config/AppConfigurationRefresh.java:11]: line 41, the api exposes a class from the impl package, `AppConfigurationStoreHealth`.

File: sdk/spring/spring-cloud-azure-appconfiguration-config-web/src/main/java/com/azure/spring/cloud/config/web/AppConfigurationWebAutoConfiguration.java
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-appconfiguration-config-web/src/main/java/com/azure/spring/cloud/config/web/AppConfigurationWebAutoConfiguration.java:45]: line 48 and 54, we can remove the `public` from the bean methods.

File: sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/config/implementation/config/AppConfigurationBootstrapConfiguration.java
Comment (saragluna) [reply sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/config/implementation/config/AppConfigurationBootstrapConfiguration.java:53]: It's okay we use this way.

File: eng/code-quality-reports/src/main/java/com/azure/tools/revapi/transforms/AzureSdkAllowedExternalApis.java
Comment (saragluna) [inline eng/code-quality-reports/src/main/java/com/azure/tools/revapi/transforms/AzureSdkAllowedExternalApis.java:86]: I just noticed that we should use `spring.cloud.azure.config` and `spring.cloud.azure.feature` as the top package name.

**Review by saragluna — CHANGES_REQUESTED** [review]

File: eng/code-quality-reports/src/main/java/com/azure/tools/revapi/transforms/AzureSdkAllowedExternalApis.java
Comment (mrm9084) [reply eng/code-quality-reports/src/main/java/com/azure/tools/revapi/transforms/AzureSdkAllowedExternalApis.java:86]: Is there a good reason for the name change at this point? And if so. We can do it in a second PR, before GA.

File: sdk/appconfiguration/tests.yml
Comment (alzimmermsft) [reply sdk/appconfiguration/tests.yml:4]: This should be kept and the `Artifacts` that were removed should instead be replaced with

```
Artifacts:
  - name: azure-data-appconfiguration
     groupId: com.azure
     safeName: azuredataappconfiguration
```

File: sdk/appconfiguration/tests.yml
Comment (alzimmermsft) [inline sdk/appconfiguration/tests.yml:15]: `TestGoals` and `TestOptions` can be deleted

File: sdk/spring/spring-cloud-azure-actuator/src/main/java/com/azure/spring/cloud/actuator/appconfiguration/AppConfigurationConfigHealthIndicator.java
Comment (mrm9084) [reply sdk/spring/spring-cloud-azure-actuator/src/main/java/com/azure/spring/cloud/actuator/appconfiguration/AppConfigurationConfigHealthIndicator.java:13]: I don't think we can as then `spring-cloud-azure-actuator-autoconfigure` would be trying to access a private class of another package.

Comment (saragluna) [reply sdk/spring/spring-cloud-azure-actuator/src/main/java/com/azure/spring/cloud/actuator/appconfiguration/AppConfigurationConfigHealthIndicator.java:13]: How about moving it to the implementation package then.

File: eng/code-quality-reports/src/main/java/com/azure/tools/revapi/transforms/AzureSdkAllowedExternalApis.java
Comment (saragluna) [reply eng/code-quality-reports/src/main/java/com/azure/tools/revapi/transforms/AzureSdkAllowedExternalApis.java:86]: Oh, I was wrong about that, `com.azure.spring.cloud` is correct.

File: sdk/spring/spring-cloud-azure-actuator/src/main/java/com/azure/spring/cloud/actuator/appconfiguration/AppConfigurationConfigHealthIndicator.java
Comment (mrm9084) [reply sdk/spring/spring-cloud-azure-actuator/src/main/java/com/azure/spring/cloud/actuator/appconfiguration/AppConfigurationConfigHealthIndicator.java:13]: I thought we didn't access implementation code of other packages?

**Pull request merged**
